docs/spec/spec.md

```diff
@@ -0,0 +1,564 @@
+# Specification: Ballerina MQTT Library
+
+_Owners_: @shafreenAnfar @dilanSachi    
+_Reviewers_: @shafreenAnfar  
+_Created_: 2023/08/15    
+_Updated_: 2023/08/15   
+_Edition_: Swan Lake
+
+## Introduction
+This is the specification for the MQTT standard library of [Ballerina language](https://ballerina.io/), which can send and receive messages 
+by connecting to a MQTT broker.
+
+The MQTT library specification has evolved and may continue to evolve in the future. The released versions of the 
+specification can be found under the relevant GitHub tag.
+
+If you have any feedback or suggestions about the library, start a discussion via a [GitHub issue](https://github.com/ballerina-platform/ballerina-standard-library/issues) or in the 
+[Discord server](https://discord.gg/ballerinalang). Based on the outcome of the discussion, the specification and implementation can be updated. Community 
+feedback is always welcome. Any accepted proposal, which affects the specification is stored under `/docs/proposals`. Proposals 
+under discussion can be found with the label `type/proposal` in GitHub.
+
+The conforming implementation of the specification is released to Ballerina central. Any deviation from the specification 
+is considered a bug.
+
+## Contents
+1. [Overview](#1-overview)
+2. [Configurations](#2-configurations)
+    *  2.1. [Security Configurations](#21-security-configurations)
+    *  2.2. [ConnectionConfiguration](#22-connectionconfiguration)
+    *  2.3. [Message](#23-message)
+    *  2.4. [DeliveryToken](#24-deliverytoken)
+    *  2.5. [Subscription](#25-subscription)
+3. [Client](#3-client)
+    *  3.1. [Configurations](#31-configurations)
+    *  3.2. [Initialization](#32-initialization)
+        *    3.2.1. [Insecure Client](#321-insecure-client)
+        *    3.2.2. [Secure Client](#322-secure-client)
+    *  3.3. [Functions](#33-functions)
+4. [Subscriber](#4-subscriber)
+    *  4.1. [Configurations](#41-configurations)
+    *  4.2. [Initialization](#42-initialization)
+        *  4.2.1. [Insecure Listener](#421-insecure-listener)
+        *  4.2.2. [Secure Listener](#422-secure-listener)
+    *  4.3. [Usage](#43-usage)
+    *  4.4. [Caller](#44-caller)
+5. [Samples](#5-samples)
```

Comment:
I don't think adding a separate section for samples is a good idea. What happens if you want to add another section to the spec? It will be a nightmare to edit BBEs, other links, and whatnot (by experience), or you may have to add a section after the `Samples` section, which would be weird. 
I'd suggest adding the samples within the sections where applicable. (Check other specs like HTTP and GraphQL for reference).

Reply:
Distributed the examples into client and listener